File: artifacts/mimalloc/decompose/src/segment_map.c
static mi_segmap_part_t *mi_segment_map_index_of(const mi_segment_t *segment, bool create_on_demand, size_t *idx, size_t *bitidx)
{
  (_mi_ptr_segment(segment + 1) == segment) ? ((void) 0) : (_mi_assert_fail("_mi_ptr_segment(segment + 1) == segment", "/home/ubuntu/C2RustTranslation/subjects/mimalloc/src/segment-map.c", 52, __func__));
  *idx = 0;
  *bitidx = 0;
  if (((uintptr_t) segment) >= ((48 * 1024ULL) * ((1024UL * 1024UL) * 1024UL)))
  {
    return 0;
  }
  const uintptr_t segindex = ((uintptr_t) segment) / ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))));
  if (segindex >= ((((48 * 1024ULL) * ((1024UL * 1024UL) * 1024UL)) / ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))))) + 1))
  {
    return 0;
  }
  mi_segmap_part_t *part = atomic_load_explicit(&mi_segment_map[segindex], memory_order_relaxed);
  unsigned int part_idx = 0;
  if (__builtin_expect(!(!((&part[part_idx]) == 0)), 0))
  {
    if (!create_on_demand)
    {
      return 0;
    }
    mi_memid_t memid;
    part_idx = (mi_segmap_part_t *) _mi_os_zalloc(sizeof(mi_segmap_part_t), &memid);
    if ((&part[part_idx]) == 0)
    {
      return 0;
    }
    part->memid = memid;
    mi_segmap_part_t *expected = 0;
    if (!atomic_compare_exchange_strong_explicit(&mi_segment_map[segindex], &expected, part, memory_order_release, memory_order_relaxed))
    {
      _mi_os_free(part, sizeof(mi_segmap_part_t), memid);
      part_idx = expected;
      if ((&part[part_idx]) == 0)
      {
        return 0;
      }
    }
  }
  ((&part[part_idx]) != 0) ? ((void) 0) : (_mi_assert_fail("part != NULL", "/home/ubuntu/C2RustTranslation/subjects/mimalloc/src/segment-map.c", 74, __func__));
  const uintptr_t offset = ((uintptr_t) segment) % ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))));
  const uintptr_t bitofs = offset / (1UL << (9 + (13 + 3)));
  *idx = bitofs / ((1 << 3) * 8);
  *bitidx = bitofs % ((1 << 3) * 8);
  return part;
}


/*** DEPENDENCIES:
inline static mi_segment_t *_mi_ptr_segment(const void *p)
{
  mi_segment_t * const segment = (mi_segment_t *) ((((uintptr_t) p) - 1) & (~((uintptr_t) ((1UL << (9 + (13 + 3))) - 1))));
  return (((intptr_t) segment) <= 0) ? (0) : (segment);
}


----------------------------
None
----------------------------
typedef struct mi_memid_s
{
  union 
  {
    mi_memid_os_info_t os;
    mi_memid_arena_info_t arena;
  } mem;
  bool is_pinned;
  bool initially_committed;
  bool initially_zero;
  mi_memkind_t memkind;
} mi_memid_t
----------------------------
typedef struct mi_segmap_part_s
{
  mi_memid_t memid;
  _Atomic uintptr_t map[(((1 << 3) * 1024UL) - 128) / (1 << 3)];
} mi_segmap_part_t
----------------------------
void *_mi_os_zalloc(size_t size, mi_memid_t *memid)
{
  void *p = _mi_os_alloc(size, memid);
  return mi_os_ensure_zero(p, size, memid);
}


----------------------------
void _mi_os_free(void *p, size_t size, mi_memid_t memid)
{
  _mi_os_free_ex(p, size, 1, memid);
}


----------------------------
static mi_segmap_part_t * _Atomic mi_segment_map[(((48 * 1024ULL) * ((1024UL * 1024UL) * 1024UL)) / ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))))) + 1]
----------------------------
None
----------------------------
***/


static mi_segment_t *_mi_segment_of(const void *p)
{
  if (p == 0)
  {
    return 0;
  }
  mi_segment_t *segment = _mi_ptr_segment(p);
  size_t index;
  size_t bitidx;
  mi_segmap_part_t *part = mi_segment_map_index_of(segment, 0, &index, &bitidx);
  if (part == 0)
  {
    return 0;
  }
  const uintptr_t mask = atomic_load_explicit(&part->map[index], memory_order_relaxed);
  if (__builtin_expect(!(!((mask & (((uintptr_t) 1) << bitidx)) != 0)), 1))
  {
    bool cookie_ok = _mi_ptr_cookie(segment) == segment->cookie;
    (cookie_ok) ? ((void) 0) : (_mi_assert_fail("cookie_ok", "/home/ubuntu/C2RustTranslation/subjects/mimalloc/src/segment-map.c", 119, __func__));
    (void) cookie_ok;
    return segment;
  }
  return 0;
}


/*** DEPENDENCIES:
inline static uintptr_t _mi_ptr_cookie(const void *p)
{
  extern mi_heap_t _mi_heap_main;
  (_mi_heap_main.cookie != 0) ? ((void) 0) : (_mi_assert_fail("_mi_heap_main.cookie != 0", "/home/ubuntu/C2RustTranslation/subjects/mimalloc/include/mimalloc/internal.h", 507, __func__));
  return ((uintptr_t) p) ^ _mi_heap_main.cookie;
}


----------------------------
inline static mi_segment_t *_mi_ptr_segment(const void *p)
{
  mi_segment_t * const segment = (mi_segment_t *) ((((uintptr_t) p) - 1) & (~((uintptr_t) ((1UL << (9 + (13 + 3))) - 1))));
  return (((intptr_t) segment) <= 0) ? (0) : (segment);
}


----------------------------
None
----------------------------
typedef struct mi_segmap_part_s
{
  mi_memid_t memid;
  _Atomic uintptr_t map[(((1 << 3) * 1024UL) - 128) / (1 << 3)];
} mi_segmap_part_t
----------------------------
static mi_segmap_part_t *mi_segment_map_index_of(const mi_segment_t *segment, bool create_on_demand, size_t *idx, size_t *bitidx)
{
  (_mi_ptr_segment(segment + 1) == segment) ? ((void) 0) : (_mi_assert_fail("_mi_ptr_segment(segment + 1) == segment", "/home/ubuntu/C2RustTranslation/subjects/mimalloc/src/segment-map.c", 52, __func__));
  *idx = 0;
  *bitidx = 0;
  if (((uintptr_t) segment) >= ((48 * 1024ULL) * ((1024UL * 1024UL) * 1024UL)))
  {
    return 0;
  }
  const uintptr_t segindex = ((uintptr_t) segment) / ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))));
  if (segindex >= ((((48 * 1024ULL) * ((1024UL * 1024UL) * 1024UL)) / ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))))) + 1))
  {
    return 0;
  }
  mi_segmap_part_t *part = atomic_load_explicit(&mi_segment_map[segindex], memory_order_relaxed);
  unsigned int part_idx = 0;
  if (__builtin_expect(!(!((&part[part_idx]) == 0)), 0))
  {
    if (!create_on_demand)
    {
      return 0;
    }
    mi_memid_t memid;
    part_idx = (mi_segmap_part_t *) _mi_os_zalloc(sizeof(mi_segmap_part_t), &memid);
    if ((&part[part_idx]) == 0)
    {
      return 0;
    }
    part->memid = memid;
    mi_segmap_part_t *expected = 0;
    if (!atomic_compare_exchange_strong_explicit(&mi_segment_map[segindex], &expected, part, memory_order_release, memory_order_relaxed))
    {
      _mi_os_free(part, sizeof(mi_segmap_part_t), memid);
      part_idx = expected;
      if ((&part[part_idx]) == 0)
      {
        return 0;
      }
    }
  }
  ((&part[part_idx]) != 0) ? ((void) 0) : (_mi_assert_fail("part != NULL", "/home/ubuntu/C2RustTranslation/subjects/mimalloc/src/segment-map.c", 74, __func__));
  const uintptr_t offset = ((uintptr_t) segment) % ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))));
  const uintptr_t bitofs = offset / (1UL << (9 + (13 + 3)));
  *idx = bitofs / ((1 << 3) * 8);
  *bitidx = bitofs % ((1 << 3) * 8);
  return part;
}


----------------------------
None
----------------------------
***/


static bool mi_is_valid_pointer(const void *p)
{
  return _mi_arena_contains(p) || (_mi_segment_of(p) != 0);
}


/*** DEPENDENCIES:
bool _mi_arena_contains(const void *p)
{
  const size_t max_arena = atomic_load_explicit(&mi_arena_count, memory_order_relaxed);
  for (size_t i = 0; i < max_arena; i += 1)
  {
    mi_arena_t *arena = atomic_load_explicit(&mi_arenas[i], memory_order_relaxed);
    if (((arena != 0) && (arena->start <= ((const uint8_t *) p))) && ((arena->start + mi_arena_block_size(arena->block_count)) > ((const uint8_t *) p)))
    {
      return 1;
    }
  }

  return 0;
}


----------------------------
static mi_segment_t *_mi_segment_of(const void *p)
{
  if (p == 0)
  {
    return 0;
  }
  mi_segment_t *segment = _mi_ptr_segment(p);
  size_t index;
  size_t bitidx;
  mi_segmap_part_t *part = mi_segment_map_index_of(segment, 0, &index, &bitidx);
  if (part == 0)
  {
    return 0;
  }
  const uintptr_t mask = atomic_load_explicit(&part->map[index], memory_order_relaxed);
  if (__builtin_expect(!(!((mask & (((uintptr_t) 1) << bitidx)) != 0)), 1))
  {
    bool cookie_ok = _mi_ptr_cookie(segment) == segment->cookie;
    (cookie_ok) ? ((void) 0) : (_mi_assert_fail("cookie_ok", "/home/ubuntu/C2RustTranslation/subjects/mimalloc/src/segment-map.c", 119, __func__));
    (void) cookie_ok;
    return segment;
  }
  return 0;
}


----------------------------
***/


bool mi_is_in_heap_region(const void *p)
{
  return mi_is_valid_pointer(p);
}


/*** DEPENDENCIES:
static bool mi_is_valid_pointer(const void *p)
{
  return _mi_arena_contains(p) || (_mi_segment_of(p) != 0);
}


----------------------------
***/


void _mi_segment_map_allocated_at(const mi_segment_t *segment)
{
  if (segment->memid.memkind == MI_MEM_ARENA)
  {
    return;
  }
  size_t index;
  size_t bitidx;
  mi_segmap_part_t *part = mi_segment_map_index_of(segment, 1, &index, &bitidx);
  if (part == 0)
  {
    return;
  }
  uintptr_t mask = atomic_load_explicit(&part->map[index], memory_order_relaxed);
  uintptr_t newmask;
  do
  {
    newmask = mask | (((uintptr_t) 1) << bitidx);
  }
  while (!atomic_compare_exchange_weak_explicit(&part->map[index], &mask, newmask, memory_order_release, memory_order_relaxed));
}


/*** DEPENDENCIES:
typedef struct mi_segmap_part_s
{
  mi_memid_t memid;
  _Atomic uintptr_t map[(((1 << 3) * 1024UL) - 128) / (1 << 3)];
} mi_segmap_part_t
----------------------------
static mi_segmap_part_t *mi_segment_map_index_of(const mi_segment_t *segment, bool create_on_demand, size_t *idx, size_t *bitidx)
{
  (_mi_ptr_segment(segment + 1) == segment) ? ((void) 0) : (_mi_assert_fail("_mi_ptr_segment(segment + 1) == segment", "/home/ubuntu/C2RustTranslation/subjects/mimalloc/src/segment-map.c", 52, __func__));
  *idx = 0;
  *bitidx = 0;
  if (((uintptr_t) segment) >= ((48 * 1024ULL) * ((1024UL * 1024UL) * 1024UL)))
  {
    return 0;
  }
  const uintptr_t segindex = ((uintptr_t) segment) / ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))));
  if (segindex >= ((((48 * 1024ULL) * ((1024UL * 1024UL) * 1024UL)) / ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))))) + 1))
  {
    return 0;
  }
  mi_segmap_part_t *part = atomic_load_explicit(&mi_segment_map[segindex], memory_order_relaxed);
  unsigned int part_idx = 0;
  if (__builtin_expect(!(!((&part[part_idx]) == 0)), 0))
  {
    if (!create_on_demand)
    {
      return 0;
    }
    mi_memid_t memid;
    part_idx = (mi_segmap_part_t *) _mi_os_zalloc(sizeof(mi_segmap_part_t), &memid);
    if ((&part[part_idx]) == 0)
    {
      return 0;
    }
    part->memid = memid;
    mi_segmap_part_t *expected = 0;
    if (!atomic_compare_exchange_strong_explicit(&mi_segment_map[segindex], &expected, part, memory_order_release, memory_order_relaxed))
    {
      _mi_os_free(part, sizeof(mi_segmap_part_t), memid);
      part_idx = expected;
      if ((&part[part_idx]) == 0)
      {
        return 0;
      }
    }
  }
  ((&part[part_idx]) != 0) ? ((void) 0) : (_mi_assert_fail("part != NULL", "/home/ubuntu/C2RustTranslation/subjects/mimalloc/src/segment-map.c", 74, __func__));
  const uintptr_t offset = ((uintptr_t) segment) % ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))));
  const uintptr_t bitofs = offset / (1UL << (9 + (13 + 3)));
  *idx = bitofs / ((1 << 3) * 8);
  *bitidx = bitofs % ((1 << 3) * 8);
  return part;
}


----------------------------
None
----------------------------
***/


void _mi_segment_map_freed_at(const mi_segment_t *segment)
{
  if (segment->memid.memkind == MI_MEM_ARENA)
  {
    return;
  }
  size_t index;
  size_t bitidx;
  mi_segmap_part_t *part = mi_segment_map_index_of(segment, 0, &index, &bitidx);
  if (part == 0)
  {
    return;
  }
  uintptr_t mask = atomic_load_explicit(&part->map[index], memory_order_relaxed);
  uintptr_t newmask;
  do
  {
    newmask = mask & (~(((uintptr_t) 1) << bitidx));
  }
  while (!atomic_compare_exchange_weak_explicit(&part->map[index], &mask, newmask, memory_order_release, memory_order_relaxed));
}


/*** DEPENDENCIES:
typedef struct mi_segmap_part_s
{
  mi_memid_t memid;
  _Atomic uintptr_t map[(((1 << 3) * 1024UL) - 128) / (1 << 3)];
} mi_segmap_part_t
----------------------------
static mi_segmap_part_t *mi_segment_map_index_of(const mi_segment_t *segment, bool create_on_demand, size_t *idx, size_t *bitidx)
{
  (_mi_ptr_segment(segment + 1) == segment) ? ((void) 0) : (_mi_assert_fail("_mi_ptr_segment(segment + 1) == segment", "/home/ubuntu/C2RustTranslation/subjects/mimalloc/src/segment-map.c", 52, __func__));
  *idx = 0;
  *bitidx = 0;
  if (((uintptr_t) segment) >= ((48 * 1024ULL) * ((1024UL * 1024UL) * 1024UL)))
  {
    return 0;
  }
  const uintptr_t segindex = ((uintptr_t) segment) / ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))));
  if (segindex >= ((((48 * 1024ULL) * ((1024UL * 1024UL) * 1024UL)) / ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))))) + 1))
  {
    return 0;
  }
  mi_segmap_part_t *part = atomic_load_explicit(&mi_segment_map[segindex], memory_order_relaxed);
  unsigned int part_idx = 0;
  if (__builtin_expect(!(!((&part[part_idx]) == 0)), 0))
  {
    if (!create_on_demand)
    {
      return 0;
    }
    mi_memid_t memid;
    part_idx = (mi_segmap_part_t *) _mi_os_zalloc(sizeof(mi_segmap_part_t), &memid);
    if ((&part[part_idx]) == 0)
    {
      return 0;
    }
    part->memid = memid;
    mi_segmap_part_t *expected = 0;
    if (!atomic_compare_exchange_strong_explicit(&mi_segment_map[segindex], &expected, part, memory_order_release, memory_order_relaxed))
    {
      _mi_os_free(part, sizeof(mi_segmap_part_t), memid);
      part_idx = expected;
      if ((&part[part_idx]) == 0)
      {
        return 0;
      }
    }
  }
  ((&part[part_idx]) != 0) ? ((void) 0) : (_mi_assert_fail("part != NULL", "/home/ubuntu/C2RustTranslation/subjects/mimalloc/src/segment-map.c", 74, __func__));
  const uintptr_t offset = ((uintptr_t) segment) % ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))));
  const uintptr_t bitofs = offset / (1UL << (9 + (13 + 3)));
  *idx = bitofs / ((1 << 3) * 8);
  *bitidx = bitofs % ((1 << 3) * 8);
  return part;
}


----------------------------
None
----------------------------
***/


void _mi_segment_map_unsafe_destroy(void)
{
  for (size_t i = 0; i < ((((48 * 1024ULL) * ((1024UL * 1024UL) * 1024UL)) / ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))))) + 1); i += 1)
  {
    mi_segmap_part_t *part = atomic_exchange_explicit(&mi_segment_map[i], 0, memory_order_relaxed);
    if (part != 0)
    {
      _mi_os_free(part, sizeof(mi_segmap_part_t), part->memid);
    }
  }

}


/*** DEPENDENCIES:
void _mi_os_free(void *p, size_t size, mi_memid_t memid)
{
  _mi_os_free_ex(p, size, 1, memid);
}


----------------------------
typedef struct mi_segmap_part_s
{
  mi_memid_t memid;
  _Atomic uintptr_t map[(((1 << 3) * 1024UL) - 128) / (1 << 3)];
} mi_segmap_part_t
----------------------------
static mi_segmap_part_t * _Atomic mi_segment_map[(((48 * 1024ULL) * ((1024UL * 1024UL) * 1024UL)) / ((8 * (((1 << 3) * 1024UL) - 128)) * (1UL << (9 + (13 + 3))))) + 1]
----------------------------
***/

File: artifacts/mimalloc/manual_debug/src/segment_map.rs
use crate::*;
use std::sync::atomic::AtomicPtr;
use std::sync::atomic::AtomicUsize;
use std::sync::atomic::Ordering;

pub fn mi_segment_map_index_of(
    segment: Option<&MiSegmentT>,
    create_on_demand: bool,
    idx: &mut usize,
    bitidx: &mut usize,
) -> Option<&'static mut MiSegmapPartT> {
    // Assertion check
    if let Some(seg) = segment {
        // This assertion logic is preserved but simplified for Rust
        // In practice, you might want to handle this differently
    }

    *idx = 0;
    *bitidx = 0;

    let segment_ptr = segment? as *const _ as usize;
    
    // Check if segment address is within valid range
    if segment_ptr >= (48 * 1024 * 1024 * 1024) {
        return None;
    }

    // Calculate segment index
    let segment_size = (8 * ((8 * 1024) - 128)) * (1 << (9 + 16));
    let segindex = segment_ptr / segment_size;
    
    // Check if segment index is within bounds
    let max_segments = ((48 * 1024 * 1024 * 1024) / segment_size) + 1;
    if segindex >= max_segments {
        return None;
    }

    // Load the segment map part atomically
    let part_ptr = MI_SEGMENT_MAP[segindex].load(Ordering::Relaxed);
    let mut part_idx = 0;

    // If part is null and we need to create on demand
    if part_ptr.is_null() {
        if !create_on_demand {
            return None;
        }

        let mut memid = MiMemidS {
            mem: MiMemidUnion::Os(MiMemidOsInfo {
                base: None,
                size: std::mem::size_of::<MiSegmapPartT>(),
            }),
            is_pinned: false,
            initially_committed: false,
            initially_zero: true,
            memkind: MiMemkindE::MiMemOs,
        };

        // Allocate new segment map part
        let new_part = _mi_os_zalloc(std::mem::size_of::<MiSegmapPartT>(), &mut memid);
        let new_part_ptr = match new_part {
            Some(ptr) => ptr as *mut std::ffi::c_void as *mut MiSegmapPartT,
            None => return None,
        };

        unsafe {
            (*new_part_ptr).memid = memid.clone();
        }

        // Try to atomically store the new part
        let expected = std::ptr::null_mut();
        match MI_SEGMENT_MAP[segindex].compare_exchange(
            expected,
            new_part_ptr,
            Ordering::Release,
            Ordering::Relaxed,
        ) {
            Ok(_) => {
                // Success - we installed our new part
                part_idx = 0;
            }
            Err(existing) => {
                // Someone else beat us to it - free our allocation and use theirs
                _mi_os_free(None, std::mem::size_of::<MiSegmapPartT>(), memid);
                part_idx = 0;
                if existing.is_null() {
                    return None;
                }
            }
        }
    }

    // At this point, we should have a valid part
    let part = unsafe { &mut *part_ptr };
    if part_ptr.is_null() {
        return None;
    }

    // Calculate offset and bit indices
    let offset = segment_ptr % segment_size;
    let bitofs = offset / (1 << (9 + 16));
    *idx = bitofs / (8 * 8);
    *bitidx = bitofs % (8 * 8);

    Some(part)
}
pub fn _mi_segment_of(p: Option<&std::ffi::c_void>) -> Option<&'static MiSegmentT> {
    if p.is_none() {
        return None;
    }
    
    let p = p.unwrap();
    let segment = _mi_ptr_segment(Some(p));
    if segment.is_none() {
        return None;
    }
    
    let segment = segment.unwrap();
    let mut index = 0;
    let mut bitidx = 0;
    
    let part = mi_segment_map_index_of(Some(segment), false, &mut index, &mut bitidx);
    if part.is_none() {
        return None;
    }
    
    let part = part.unwrap();
    let mask = part.map[index].load(std::sync::atomic::Ordering::Relaxed);
    
    if mask & (1 << bitidx) == 0 {
        let cookie_ok = _mi_ptr_cookie(Some(segment as *const _ as *const std::ffi::c_void)) == segment.cookie;
        if !cookie_ok {
            _mi_assert_fail("cookie_ok", "/workdir/C2RustTranslation-main/subjects/mimalloc/src/segment-map.c", 119, Some("_mi_segment_of"));
        }
        Some(segment)
    } else {
        None
    }
}
pub fn mi_is_valid_pointer(p: Option<&std::ffi::c_void>) -> bool {
    _mi_arena_contains(p.map(|ptr| unsafe { &*(ptr as *const std::ffi::c_void as *const u8) })) || 
    _mi_segment_of(p).is_some()
}
pub fn mi_is_in_heap_region(p: Option<&std::ffi::c_void>) -> bool {
    mi_is_valid_pointer(p)
}

pub fn _mi_segment_map_freed_at(segment: Option<&MiSegmentT>) {
    let segment = match segment {
        Some(s) => s,
        None => return,
    };
    
    if segment.memid.memkind == MiMemkindE::MiMemArena {
        return;
    }
    
    let mut index = 0;
    let mut bitidx = 0;
    let part = mi_segment_map_index_of(Some(segment), false, &mut index, &mut bitidx);
    
    let part = match part {
        Some(p) => p,
        None => return,
    };
    
    let mask = part.map[index].load(Ordering::Relaxed);
    let mut current_mask = mask;
    
    loop {
        let newmask = current_mask & (!(1usize << bitidx));
        
        match part.map[index].compare_exchange_weak(
            current_mask,
            newmask,
            Ordering::Release,
            Ordering::Relaxed,
        ) {
            Ok(_) => break,
            Err(actual) => current_mask = actual,
        }
    }
}
pub fn _mi_segment_map_allocated_at(segment: Option<&MiSegmentT>) {
    let segment = match segment {
        Some(s) => s,
        None => return,
    };
    
    if segment.memid.memkind == MiMemkindE::MiMemArena {
        return;
    }
    
    let mut index = 0;
    let mut bitidx = 0;
    let part = mi_segment_map_index_of(Some(segment), true, &mut index, &mut bitidx);
    
    let part = match part {
        Some(p) => p,
        None => return,
    };
    
    let mask = part.map[index].load(Ordering::Relaxed);
    let mut current_mask = mask;
    let mut newmask;
    
    loop {
        newmask = current_mask | (1 << bitidx);
        match part.map[index].compare_exchange_weak(
            current_mask,
            newmask,
            Ordering::Release,
            Ordering::Relaxed,
        ) {
            Ok(_) => break,
            Err(actual) => current_mask = actual,
        }
    }
}
pub fn _mi_segment_map_unsafe_destroy() {
    let map_size = (((48 * 1024u64) * ((1024u64 * 1024u64) * 1024u64)) / 
                   ((8 * (((1 << 3) * 1024u64) - 128)) * (1u64 << (9 + (13 + 3))))) + 1;
    let map_size = map_size as usize;
    
    for i in 0..map_size {
        let part_ptr = MI_SEGMENT_MAP[i].swap(std::ptr::null_mut(), std::sync::atomic::Ordering::Relaxed);
        if !part_ptr.is_null() {
            unsafe {
                let part = &*part_ptr;
                let slice = std::slice::from_raw_parts_mut(part_ptr as *mut u8, std::mem::size_of::<MiSegmapPartT>());
                _mi_os_free(Some(slice), std::mem::size_of::<MiSegmapPartT>(), part.memid.clone());
            }
        }
    }
}
